docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,244 @@
+defined as a `StateT` with
+[Eval](https://typelevel.org/cats/datatypes/eval.html) as the effect
+`F`.  Therefore, `StateT` exposes the same methods of
+[State](https://typelevel.org/cats/datatypes/state.html), such as:
+`modify`, `get` and `set`.  Plus additional methods, that handles
+effectful computations, eg: `modifyF`, `setF` and `liftF`.
+
+## Example: Table Reservation System
+
+In this example, we are going to model a *Table Reservation System*. To
+do so, we need to keep track of the tables and the current
+reservations for each of them. The end-user can then try to insert a
+booking for a specific table and time. If such a table is available,
+then the booking is placed and the state is updated, otherwise, an
+error is returned.
+
+To simplify the logic, for each reservation we will just consider a
+single `LocalTime` starting at the beginning of the hour.
+
+Let's start with defining the type synonyms for the state of the
+program and the effect type:
+
+```scala mdoc:silent
+import cats.data.{StateT, NonEmptyList}
+import cats.implicits._
+import java.time.LocalTime
+
+type TRSState = Map[(Int, LocalTime), String]
+type EffectType[A] = Either[Throwable, A]
+```
+
+
+`TRSState` stands for *Table Reservation System State* and it will
+organize the reservations in a `Map` with:
+- `(Int, LocalTime)` as key, representing the table number and hour
+respectively.
+- `String` as value, representing the reservation name.
+
+As effect type, we will just use an `Either[Throwable, A]` because each
+booking insertion can fail.
+
+Now, we need to implement/define:
+- An initial state, that will be just an empty `Map`.
+- A custom `Throwable` to be used in case of an error.
+- The logic for the booking insertion. We can take advantage of the
+method `modifyF`.
+
+In addition, we can implement a simple function that will evaluate a
+`NonEmptyList` of reservations, inserting them one by one.
+
+```scala mdoc:silent
+object TableReservationSystem {
+
+  final case class AlreadyReservedTable(
+      tableNumber: Int,
+      time: LocalTime,
+      reservation: String
+  ) extends Throwable(
+        s"$reservation cannot be added because table number $tableNumber is already reserved for the $time"
+      )
+
+  val emptyReservationSystem: TRSState = Map.empty
+
+  def insertBooking(
+      tableNumber: Int,
+      time: LocalTime,
+      name: String
+  ): StateT[EffectType, TRSState, Unit] =
+    StateT.modifyF[EffectType, TRSState]((currentReservations: TRSState) =>
+      currentReservations
+        .get((tableNumber, time))
+        .fold[EffectType[TRSState]](Right(currentReservations + ((tableNumber, time) -> name)))(_ =>
+          Left(AlreadyReservedTable(tableNumber, time, name))
+        )
+    )
+
+  def evalBookings(
+      bookings: NonEmptyList[(Int, LocalTime, String)]
+  ): EffectType[TRSState] =
+    bookings
+      .map { case (tn, t, n) =>
+        TableReservationSystem.insertBooking(tn, t, n)
+      }
+      .reduceLeft(_ *> _)
```

Comment:
I worry if this line is confusing because of all the fancy syntax.

Reply:
This is actually the line for composing the `StateT`s to a single one. I'm open to suggestions to make it less cryptic 🙌

Reply:
Sure :) maybe something like this.
```scala
.reduceLeft((acc, bookNext) => acc.productL(bookNext))
```